crates/ruff_linter/src/rules/isort/sorting.rs

```diff
@@ -1,171 +1,137 @@
-    level2: Option<u32>,
-    relative_imports_order: RelativeImportsOrder,
-) -> Ordering {
-    match (level1, level2) {
-        (None, None) => Ordering::Equal,
-        (None, Some(_)) => Ordering::Less,
-        (Some(_), None) => Ordering::Greater,
-        (Some(level1), Some(level2)) => match relative_imports_order {
-            RelativeImportsOrder::ClosestToFurthest => level1.cmp(&level2),
-            RelativeImportsOrder::FurthestToClosest => level2.cmp(&level1),
-        },
+#[derive(Eq, PartialEq, Debug)]
+pub(crate) struct NatOrdStr<'a>(&'a str);
+
+impl Ord for NatOrdStr<'_> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        natord::compare(self.0, other.0)
     }
 }
 
-/// Compare two `Stmt::ImportFrom` blocks.
-pub(crate) fn cmp_import_from(
-    import_from1: &ImportFromData,
-    import_from2: &ImportFromData,
-    settings: &Settings,
-) -> Ordering {
-    cmp_levels(
-        import_from1.level,
-        import_from2.level,
-        settings.relative_imports_order,
-    )
-    .then_with(|| {
-        cmp_force_to_top(
-            &import_from1.module_name(),
-            &import_from2.module_name(),
-            &settings.force_to_top,
-        )
-    })
-    .then_with(|| match (&import_from1.module, import_from2.module) {
-        (None, None) => Ordering::Equal,
-        (None, Some(_)) => Ordering::Less,
-        (Some(_), None) => Ordering::Greater,
-        (Some(module1), Some(module2)) => {
-            if settings.case_sensitive {
-                natord::compare(module1, module2)
-            } else {
-                natord::compare_ignore_case(module1, module2)
-            }
-        }
-    })
+impl PartialOrd for NatOrdStr<'_> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
 }
 
-/// Compare an import to an import-from.
-fn cmp_import_import_from(
-    import: &AliasData,
-    import_from: &ImportFromData,
+type ModuleKey<'a> = (
+    i64,
+    Option<bool>,
+    Option<NatOrdString>,
+    Option<NatOrdStr<'a>>,
+    Option<NatOrdStr<'a>>,
+    Option<MemberKey<'a>>,
+);
+
+pub(crate) fn module_key<'a>(
+    name: Option<&'a str>,
+    asname: Option<&'a str>,
+    level: Option<u32>,
+    first_alias: Option<(&'a str, Option<&'a str>)>,
     settings: &Settings,
-) -> Ordering {
-    cmp_force_to_top(
-        import.name,
-        &import_from.module_name(),
-        &settings.force_to_top,
+) -> ModuleKey<'a> {
+    let level = level
+        .map(i64::from)
+        .map(|l| match settings.relative_imports_order {
+            RelativeImportsOrder::ClosestToFurthest => l,
+            RelativeImportsOrder::FurthestToClosest => -l,
+        })
+        .unwrap_or_default();
+    let force_to_top = name.map(|name| !settings.force_to_top.contains(name)); // `false` < `true` so we get forced to top first
+    let maybe_lower_case_name = name
+        .and_then(|name| (!settings.case_sensitive).then_some(NatOrdString(name.to_lowercase())));
```

Comment:
Can we avoid allocating here (`name.to_lowercase()`) in the event that the string is already lowercase?

Reply:
I guess to do that we would need to use a `Cow`. What we could do is have a single `NatOrdStr` struct like this:
```rust
use std::borrow::Cow;

struct NatOrdStr<'a>(Cow<'a, str>);
```
whose contents are either owned or borrowed.

I haven't played around with `Cow`s much, so I don't know if the overhead is low enough to make it worthwhile to avoid that allocation in case the string is already lowercase, what do you think? If you think it is I'll gladly tweak my implementation :) but in that case I think it's better to use that `Cow`-based struct everywhere rather than just for `maybe_lower_case_name`.

Reply:
I think it's worth it given that these are gonna lower-cased in the majority of cases.

Reply:
Just pushed the modification, I had it ready just in case :)